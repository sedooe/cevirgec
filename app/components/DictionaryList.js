import React, { Component, PropTypes } from 'react';
import { Button, Checkbox, List, Popup } from 'semantic-ui-react';
import DictionaryModal from './DictionaryModal';
import tr from '../utils/Translation';

const propFunctionProxy = (prop, dictionary) => {
  prop(dictionary);
}

const DictionaryList = ({dictionaries, onEdit, onDelete, onCheckboxToggle}) => (
  <List divided relaxed>
    {Object.getOwnPropertyNames(dictionaries).map((key) => {
      let dictionary = dictionaries[key];
      return (
        <List.Item key={'dictionary-' + key}>
          <List.Content floated='right'>
            <Popup  trigger={<Button icon='add' />} content={tr('Add a new word to this dictionary')}/>
            <Popup trigger={<Button icon='print' />} content={tr('Print')}/>
            <Popup trigger={<Button icon='unhide' />} content={tr('View dictionary content')}/>
            <Popup trigger={<Button icon='edit' onClick={propFunctionProxy.bind(null, onEdit, dictionary)} />} content={tr('Edit')}/>
            <Popup trigger={<Button icon='trash' onClick={propFunctionProxy.bind(null, onDelete, dictionary)} />} content={tr('Delete')}/>
          </List.Content>
          <List.Content floated='left'>
            <Checkbox toggle checked={dictionary.active} onChange={propFunctionProxy.bind(null, onCheckboxToggle, dictionary.id)} />
          </List.Content>

          <List.Content>
            <List.Header>{dictionary.name}</List.Header>
            <List.Description>{dictionary.sourceLanguage} => {dictionary.targetLanguage} | {dictionary.numberOfDefinitions} definitions</List.Description>
          </List.Content>
        </List.Item>
      )
    })}
  </List>
)

DictionaryList.propTypes = {
  dictionaries: PropTypes.object.isRequired,
  onEdit: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired,
  onCheckboxToggle: PropTypes.func.isRequired
};

export default DictionaryList
