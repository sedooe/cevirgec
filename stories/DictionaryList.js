import React, { Component } from 'react';
import { Button, Checkbox, List, Popup } from 'semantic-ui-react'
import tr from '../app/utils/Translation';
// <CheckboxToggle label={tr('In Use')} name="active" defaultChecked={dictionary.active} onChange={that.toggleUsage.bind(that, dictionary)} />

class DictionaryList extends Component {

  render() {
    let dictionaryMap = {1: {
      name: "deneme",
      sourceLanguage: "af",
      targetLanguage: "tr",
      numberOfDefinitions: "9",
      active: true
    },
    2: {
      name: "deneme",
      sourceLanguage: "za",
      targetLanguage: "tr",
      numberOfDefinitions: "90",
      active: true
    }
  };

    return (
      <List divided relaxed>
        {Object.getOwnPropertyNames(dictionaryMap).map(function (key) {
          let dictionary = dictionaryMap[key];
          return (
            <List.Item>
              <List.Content floated='right'>
                <Popup trigger={<Button icon='add' />} content={tr('Add a new word to this dictionary')}/>
                <Popup trigger={<Button icon='print' />} content={tr('Print')}/>
                <Popup trigger={<Button icon='unhide' />} content={tr('View dictionary content')}/>
                <Popup trigger={<Button icon='edit' />} content={tr('Edit')}/>
                <Popup trigger={<Button icon='trash' />} content={tr('Delete')}/>
              </List.Content>
              <List.Content floated='left'>
                <Checkbox toggle />
              </List.Content>

              <List.Content>
                <List.Header>{dictionary.name}</List.Header>
                <List.Description>{dictionary.sourceLanguage} => {dictionary.targetLanguage} | {dictionary.numberOfDefinitions} definitions</List.Description>
              </List.Content>
            </List.Item>
          )
        })}
      </List>
    );
  }
}

export default DictionaryList
