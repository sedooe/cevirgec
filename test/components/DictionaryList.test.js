import React from 'react';
import DictionaryList from '../../app/components/DictionaryList';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import { mount } from 'enzyme';
import { expect as hope } from 'chai';

function empty(){}

let dictionaries = {
  "1": {
    id: "1",
    name: "deneme",
    sourceLanguage: "af",
    targetLanguage: "tr",
    context: "sport",
    numberOfDefinitions: "9",
    active: true
  },
  "2": {
    id: "2",
    name: "sozluk",
    sourceLanguage: "za",
    targetLanguage: "tr",
    context: "technology",
    numberOfDefinitions: "90",
    active: false
  }
};

describe('DictionaryList Component', function () {

  describe('DictionaryList renders correctly [snapshots]', function () {
    it('renders correctly with data', () => {
      const component = renderer.create(
        <DictionaryList
          dictionaries={dictionaries}
          onEdit={empty}
          onDelete={empty}
          onCheckboxToggle={empty} />
      );

      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('shows empty message with no data', () => {
      const component = renderer.create(
        <DictionaryList
          dictionaries={ {} }
          onEdit={empty}
          onDelete={empty}
          onCheckboxToggle={empty} />
      );

      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('DictionaryList renders list item\'s DOM elements correctly', function () {
    it('shows empty message with no data [DOM check]', () => {
      const component = shallow(
        <DictionaryList
          dictionaries={ {} }
          onEdit={empty}
          onDelete={empty}
          onCheckboxToggle={empty} />
      );

      // Keep below as example
      hope(component.find('.info')).to.have.length(1);
      // or
      expect(component.find('.info').length).toEqual(1);
    });

    it('should show properties of listed dictionaries', () => {
      // Render a checkbox with label in the document
      const component = mount(
        <DictionaryList
          dictionaries={dictionaries}
          onEdit={empty}
          onDelete={empty}
          onCheckboxToggle={empty} />
      );

      expect(component.find('.item .header').first().text()).toEqual('deneme');

      expect(component.find('[type="checkbox"]').first().prop('checked')).toEqual(true);
      expect(component.find('[type="checkbox"]').at(1).prop('checked')).toEqual(false);
    });

    // FIXME this doesn't work
    // https://github.com/Semantic-Org/Semantic-UI-React/blob/master/src/modules/Checkbox/Checkbox.js#L148
    it.skip('that checkbox changes state after click', () => {

      const component = mount(
        <DictionaryList
          dictionaries={dictionaries}
          onEdit={empty}
          onDelete={empty}
          onCheckboxToggle={empty} />
      );

      //don't know why this doesn't work :(

      console.log(component.find('[type="checkbox"]').first().prop('checked'));
      expect(component.find('[type="checkbox"]').first().prop('checked')).toEqual(true);
      component.find('[type="checkbox"]').first().simulate('click', {});
      console.log(component.find('[type="checkbox"]').first().prop('checked'));
      expect(component.find('[type="checkbox"]').first().prop('checked')).toEqual(false);
    });

  });
});
