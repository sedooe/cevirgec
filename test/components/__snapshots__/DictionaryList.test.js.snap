exports[`DictionaryList Component DictionaryList renders correctly [snapshots] renders correctly with data 1`] = `
<div
  className="ui divided relaxed list">
  <div
    className="item"
    data-value={undefined}>
    <div
      className="right floated content">
      <button
        className="ui icon button"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        tabIndex={undefined}>
        <i
          className="add icon" />
         
      </button>
      <button
        className="ui icon button"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        tabIndex={undefined}>
        <i
          className="print icon" />
         
      </button>
      <button
        className="ui icon button"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        tabIndex={undefined}>
        <i
          className="unhide icon" />
         
      </button>
      <button
        className="ui icon button"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        tabIndex={undefined}>
        <i
          className="edit icon" />
         
      </button>
      <button
        className="ui icon button"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        tabIndex={undefined}>
        <i
          className="trash icon" />
         
      </button>
    </div>
    <div
      className="left floated content">
      <div
        className="ui checked fitted toggle checkbox"
        onChange={[Function]}
        onClick={[Function]}>
        <input
          checked={true}
          className="hidden"
          name={undefined}
          readOnly={true}
          tabIndex={0}
          type="checkbox"
          value={undefined} />
        <label />
      </div>
    </div>
    <div
      className="content">
      <div
        className="header">
        deneme
      </div>
      <div
        className="description">
        af
        => 
        tr
        | 
        9
        definitions
      </div>
    </div>
  </div>
  <div
    className="item"
    data-value={undefined}>
    <div
      className="right floated content">
      <button
        className="ui icon button"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        tabIndex={undefined}>
        <i
          className="add icon" />
         
      </button>
      <button
        className="ui icon button"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        tabIndex={undefined}>
        <i
          className="print icon" />
         
      </button>
      <button
        className="ui icon button"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        tabIndex={undefined}>
        <i
          className="unhide icon" />
         
      </button>
      <button
        className="ui icon button"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        tabIndex={undefined}>
        <i
          className="edit icon" />
         
      </button>
      <button
        className="ui icon button"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        tabIndex={undefined}>
        <i
          className="trash icon" />
         
      </button>
    </div>
    <div
      className="left floated content">
      <div
        className="ui fitted toggle checkbox"
        onChange={[Function]}
        onClick={[Function]}>
        <input
          checked={false}
          className="hidden"
          name={undefined}
          readOnly={true}
          tabIndex={0}
          type="checkbox"
          value={undefined} />
        <label />
      </div>
    </div>
    <div
      className="content">
      <div
        className="header">
        sozluk
      </div>
      <div
        className="description">
        za
        => 
        tr
        | 
        90
        definitions
      </div>
    </div>
  </div>
</div>
`;

exports[`DictionaryList Component DictionaryList renders correctly [snapshots] shows empty message with no data 1`] = `
<div
  className="ui icon message">
  <i
    className="info icon" />
  <div
    className="content">
    There is no dictionary yet.
  </div>
</div>
`;
